src/codegate/pipeline/pii/analyzer.py

```diff
@@ -0,0 +1,168 @@
+import uuid
+from typing import Any, Dict, List, Tuple
+
+from presidio_analyzer import AnalyzerEngine
+from presidio_anonymizer import AnonymizerEngine
+import structlog
+
+logger = structlog.get_logger("codegate.pii.analyzer")
+
+
+class PiiSessionStore:
+    """
+    A class to manage PII (Personally Identifiable Information) session storage.
+
+    Attributes:
+        session_id (str): The unique identifier for the session. If not provided, a new UUID
+        is generated. mappings (Dict[str, str]): A dictionary to store mappings between UUID
+        placeholders and PII.
+
+    Methods:
+        add_mapping(pii: str) -> str:
+            Adds a PII string to the session store and returns a UUID placeholder for it.
+
+        get_pii(uuid_placeholder: str) -> str:
+            Retrieves the PII string associated with the given UUID placeholder. If the placeholder
+            is not found, returns the placeholder itself.
+    """
+
+    def __init__(self, session_id: str = None):
+        self.session_id = session_id or str(uuid.uuid4())
+        self.mappings: Dict[str, str] = {}
+
+    def add_mapping(self, pii: str) -> str:
+        uuid_placeholder = f"<{str(uuid.uuid4())}>"
+        self.mappings[uuid_placeholder] = pii
+        return uuid_placeholder
+
+    def get_pii(self, uuid_placeholder: str) -> str:
+        return self.mappings.get(uuid_placeholder, uuid_placeholder)
+
+
+class PiiAnalyzer:
+    """
+    PiiAnalyzer class for analyzing and anonymizing text containing PII.
+    Methods:
+        __init__:
+            Initializes the PiiAnalyzer with a custom NLP engine configuration.
+        analyze:
+                text (str): The text to analyze for PII.
+                Tuple[str, List[Dict[str, Any]], PiiSessionStore]: The anonymized text, a list of
+                found PII details, and the session store.
+                entities (List[str]): The PII entities to analyze for.
+
+        restore_pii:
+                anonymized_text (str): The text with anonymized PII.
+                session_store (PiiSessionStore): The PiiSessionStore used for anonymization.
+                str: The text with original PII restored.
+    """
+
+    def __init__(self):
+        import os
+
+        from presidio_analyzer.nlp_engine import NlpEngineProvider
+
+        # Get the path to our custom spacy config
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+        config_path = os.path.join(current_dir, "spacy_config.yaml")
+
+        # Initialize the NLP engine with our custom configuration
+        provider = NlpEngineProvider(conf_file=config_path)
+        nlp_engine = provider.create_engine()
```

Comment:
I wonder if this should be instantiated as a singleton -- see the other comment about creating a PiiManager once.

Reply:
fixed now, all singletons